File: android/app/src/main/kotlin/com/thecityandthebike/data/model/AppError.kt
package com.thecityandthebike.data.model

sealed class AppError {
    data class Network(val cause: Throwable) : AppError()
    data class Auth(val code: Int, val message: String) : AppError()
    data class Validation(val field: String, val message: String) : AppError()
    data class Server(val code: Int, val message: String) : AppError()
    data class RateLimit(val retryAfterSeconds: Int? = null) : AppError()
    data class Unknown(val cause: Throwable) : AppError()

    val displayMessage: String get() = when (this) {
        is Network -> "Network error. Check your connection and try again."
        is Auth -> message
        is Validation -> message
        is Server -> message
        is RateLimit -> "Too many attempts. Please try again later."
        is Unknown -> "An unexpected error occurred."
    }

    /** Returns null for Network errors (handled by global network banner). */
    val displayMessageOrNull: String? get() = if (this is Network) null else displayMessage
}

File: ios/Packages/TCATBModels/Sources/TCATBModels/AppError.swift
import Foundation

public enum AppError: Error, Sendable, Equatable {
    case network(String)
    case auth(code: Int, message: String)
    case validation(field: String, message: String)
    case server(code: Int, message: String)
    case rateLimit(retryAfterSeconds: Int?)
    case unknown(String)

    public var displayMessage: String {
        switch self {
        case .network:
            return "Network error. Check your connection and try again."
        case .auth(_, let message):
            return message
        case .validation(_, let message):
            return message
        case .server(_, let message):
            return message
        case .rateLimit:
            return "Too many attempts. Please try again later."
        case .unknown:
            return "An unexpected error occurred."
        }
    }
}
